software/contrib/cvecorder.py

```diff
@@ -0,0 +1,348 @@
+        self.debug = True
+        self.CvIn = 0
+        self.bankToSave = 0
+        self.debugTest = False
+
+        self.numCVR = 5  # Number of CV recorder channels - zero based
+        self.numCVRBanks = 5  # Number of CV recording channel banks - zero based
+
+        #self.CVR = []  # CV recorder channels
+        #self.CvRecording = []  # CV recorder flags
+
+        # Load CV Recordings from a previously stored state on disk or initialize if blank
+        self.loadState()
+
+        # Test routine, pick a random bank n times and save, then load the state
+        if self.debugTest:
+            print(micropython.mem_info("level"))
+            for n in range(3000):
+                # Clear vars
+                #self.CvRecording = []
+                print(f"Running test: {n}")
+                self.ActiveBank = randint(0, self.numCVRBanks)
+                self.ActiveCvr = randint(0, self.numCVR)
+                for i in range(0, self.stepLength-1):
+                    self.CVR[self.ActiveBank][self.ActiveCvr][i] = uniform(0.0, 9.99)
+                    #print(f"[{self.ActiveBank}][{self.ActiveCvr}][{i}] = {self.CVR[self.ActiveBank][self.ActiveCvr][i]}")
+                self.bankToSave = self.ActiveBank
+                self.saveState()
+                self.loadState()
+
+        @din.handler
+        def dInput():
+            self.handleClock()
+            self.clockStep +=1
+
+        @din.handler_falling
+        def endClock():
+            self.handleClock()
+
+        @b1.handler
+        def b1Pressed():
+            # Set recording boolean to true and clear the recording buffer
+            self.CvRecording[self.ActiveCvr] = 'pending'
+            # Clear the array
+            for n in range (0, self.stepLength):
+                self.CVR[self.ActiveBank][self.ActiveCvr][n] = 0
+
+        # B2 Long press
+        @b2.handler_falling
+        def b2PressedLong():
+            # 2000ms press clears all banks
+            if ticks_diff(ticks_ms(), b2.last_pressed()) >  2000:
+                self.clearCvrs('all')
+                #self.bankToSave = self.ActiveBank
+                #self.saveState()
+                # reverse the ActiveCvr increment caused by the initial button press
+                if self.ActiveCvr > 0:
+                    self.ActiveCvr -= 1
+                else:
+                    self.ActiveCvr = self.numCVR
+            # 500ms press clears the active bank
+            elif ticks_diff(ticks_ms(), b2.last_pressed()) >  500:
+                self.clearCvrs(self.ActiveBank)
+                self.bankToSave = self.ActiveBank
+                self.saveState()
+                # reverse the ActiveCvr increment caused by the initial button press
+                if self.ActiveCvr > 0:
+                    self.ActiveCvr -= 1
+                else:
+                    self.ActiveCvr = self.numCVR
+
+        # B2 short press
+        @b2.handler
+        def b2Pressed():
```

Comment:
I feel like it might simplify the code a bit to put all the code for button 2 in the falling handler. Then you wouldn't have to do the reversal steps.

Reply:
It would simplify the code, but I did it this way by design so that the most used function of the button triggers when the button is pressed and the least used functions trigger then the button is released.